File: students-backend/src/main/java/com/github/voodu/studentrest/controller/StudentController.java
package com.github.voodu.studentrest.controller;

import com.github.voodu.studentrest.model.Student;
import com.github.voodu.studentrest.service.StudentService;
import com.github.voodu.studentrest.repository.CourseRepository;
import com.github.voodu.studentrest.repository.StudentRepository;
import com.github.voodu.studentrest.service.AppUserService;
import com.github.voodu.studentrest.service.CourseService;
import com.github.voodu.studentrest.utility.Lambda0;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

@RestController
@CrossOrigin(origins = "*", maxAge = 3600)
@RequestMapping("/public/students")
public class StudentController {

    private AppUserService appUserService;
    private StudentService studentService;

    @Autowired
    public StudentController(AppUserService appUserService, StudentService studentService) {
        this.appUserService = appUserService;
        this.studentService = studentService;
    }

    @GetMapping
    public List<Student> findAllStudents(HttpServletRequest request, HttpServletResponse response) {
        return whenAuthorized(2, request, response, () -> studentService.findAll());
    }

    @GetMapping(value = "/{id}")
    public ResponseEntity<Student> findStudent(@PathVariable long id, HttpServletRequest request, HttpServletResponse response) {
        return whenAuthorized(2, request, response, () -> {
            Student found = studentService.getOne(id);
            if (found != null) {
                return new ResponseEntity<>(found, HttpStatus.OK);
            }
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        });
    }

    @PostMapping
    public Student addStudent(@RequestBody Student student, HttpServletRequest request, HttpServletResponse response) {
        return whenAuthorized(1, request, response, () -> {
            student.setId(null);
            return studentService.save(student);
        });
    }

    @PutMapping
    public ResponseEntity<Student> putStudent(@RequestBody Student student, HttpServletRequest request, HttpServletResponse response) {
        return whenAuthorized(0, request, response, () -> { // todo not sure what will happen now... returning null as responseEntity...
            Student saved = studentService.update(student);
            if (saved != null) {
                return new ResponseEntity<>(saved, HttpStatus.OK);
            }

            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        });
    }

    @DeleteMapping(value = "/{id}")
    public ResponseEntity<Student> deleteStudent(@PathVariable("id") long id, HttpServletRequest request, HttpServletResponse response) {
        return whenAuthorized(0, request, response, () -> {
            if (studentService.deleteById(id)) {
                return new ResponseEntity<>(HttpStatus.OK);
            }

            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        });
    }

    private boolean validateAccess(HttpServletRequest request, int requiredAccess) {
        return appUserService.validateAccess(request.getHeader("Token"), requiredAccess);
    }

    private <T> T whenAuthorized(int requiredAccess, HttpServletRequest request, HttpServletResponse response, Lambda0<T> lambda) {
        if (!validateAccess(request, requiredAccess)) {
            response.setStatus(401);
            return null;
        } else {
            return lambda.action();
        }
    }
}

File: students-backend/src/main/java/com/github/voodu/studentrest/controller/StudentController.kt
package com.github.voodu.studentrest.controller

import com.github.voodu.studentrest.model.*
import com.github.voodu.studentrest.service.CourseInfoService
import com.github.voodu.studentrest.service.StudentService
import com.github.voodu.studentrest.service.AppUserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

import javax.servlet.http.HttpServletRequest

@RestController
@CrossOrigin(origins = arrayOf("*"), maxAge = 3600)
@RequestMapping("/public/students")
class StudentController @Autowired
constructor(appUserService: AppUserService, private val studentService: StudentService, private val courseInfoService: CourseInfoService): AuthorizedController(appUserService) {

    @GetMapping
    fun findAllStudents(request: HttpServletRequest) = whenAuthorized(1, request) {
        ResponseEntity(studentService.findAll(), HttpStatus.OK)
    }.let { response ->
        if(response.statusCode == HttpStatus.OK) {
            response
        } else {
            val found = studentService.findAll().map(::SimpleStudent)
            ResponseEntity(found, HttpStatus.OK)
        }

    }

    @GetMapping(value = ["/{id}"])
    fun findStudent(@PathVariable id: Long, request: HttpServletRequest) = whenAuthorized(1, request) {
            val found = studentService.getOne(id) ?: return@whenAuthorized ResponseEntity<Student>(HttpStatus.NOT_FOUND)
            ResponseEntity(found, HttpStatus.OK)
    }

    @GetMapping(value = ["/me"])
    fun findMe(request: HttpServletRequest) = whenAuthorized<Student>(2, request) {
        val myUsername = appUserService.getMyUsername(request.getHeader("Token"))
                ?: return@whenAuthorized ResponseEntity(HttpStatus.UNAUTHORIZED) //TODO throw WTF there
        val myIndexNumber: Long
        try {
            myIndexNumber = java.lang.Long.parseLong(myUsername)
        } catch (e: NumberFormatException) {
            return@whenAuthorized ResponseEntity(HttpStatus.NOT_FOUND)
        }
        return@whenAuthorized studentService.getOne(myIndexNumber)?.let { me ->
            ResponseEntity.ok(me)
        } ?: ResponseEntity(HttpStatus.NOT_FOUND)

    }

    @PostMapping
    fun addStudent(@RequestBody student: Student, request: HttpServletRequest) = whenAuthorized(0, request) {
        ResponseEntity(studentService.save(student.apply { id = null }), HttpStatus.OK)
    }

    @PutMapping
    fun putStudent(@RequestBody student: Student, request: HttpServletRequest) = whenAuthorized(0, request) {
        val saved = studentService.update(student)
        if (saved != null) {
            ResponseEntity.ok(saved)
        } else {
            ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

    @DeleteMapping(value = ["/{id}"])
    fun deleteStudent(@PathVariable("id") id: Long, request: HttpServletRequest) = whenAuthorized(0, request) {
        val deleted = studentService.deleteById(id)
        if (deleted != null) {
            ResponseEntity.ok(deleted)
        } else {
            ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

}
